File: go/internal/types/action.go
/**
 * Author: Deepankar Das
 */

package types

// ActionType enumerates the types of actions an agent can perform.
type ActionType string

const (
	ActionFileRead       ActionType = "file.read"
	ActionFileWrite      ActionType = "file.write"
	ActionFileDelete     ActionType = "file.delete"
	ActionFileMove       ActionType = "file.move"
	ActionShellExec      ActionType = "shell.exec"
	ActionNetworkRequest ActionType = "network.request"
	ActionPackageInstall ActionType = "package.install"
	ActionCredAccess        ActionType = "credential.access"
	ActionMcpInvoke         ActionType = "mcp.invoke"
	ActionInternalOrch      ActionType = "internal.orchestration"
)

// ResourceKind classifies the resource being accessed.
type ResourceKind string

const (
	ResourceFile       ResourceKind = "file"
	ResourceCommand    ResourceKind = "command"
	ResourceHost       ResourceKind = "host"
	ResourceMcpTool    ResourceKind = "mcp_tool"
	ResourceCredential ResourceKind = "credential"
	ResourceDatabase   ResourceKind = "database"
	ResourceInternal   ResourceKind = "internal"
)

// ResourceClassification tags risk attributes of a resource.
type ResourceClassification string

const (
	ClassDestructive    ResourceClassification = "destructive"
	ClassNetworkTool    ResourceClassification = "network_tool"
	ClassPackageManager ResourceClassification = "package_manager"
	ClassSensitivePath  ResourceClassification = "sensitive_path"
	ClassSafe           ResourceClassification = "safe"
	ClassPotentialExfil ResourceClassification = "potential_exfiltration"
	ClassBypassAttempt  ResourceClassification = "bypass_attempt"
)

// Actor identifies who is performing the action.
type Actor struct {
	UserID        string `json:"user_id"`
	AgentType     string `json:"agent_type"`
	AgentInstance string `json:"agent_instance"`
	SessionID     string `json:"session_id"`
}

// Environment describes the workspace context.
type Environment struct {
	Workspace      string `json:"workspace"`
	Repo           string `json:"repo"`
	Branch         string `json:"branch"`
	Tier           string `json:"tier"`
	DeploymentMode string `json:"deployment_mode"`
}

// Resource describes the target of the action.
type Resource struct {
	Kind           ResourceKind             `json:"kind"`
	Path           string                   `json:"path,omitempty"`
	Host           string                   `json:"host,omitempty"`
	Value          string                   `json:"value,omitempty"`
	Classification []ResourceClassification `json:"classification"`
}

// ActionDetail describes what the agent is trying to do.
type ActionDetail struct {
	Type            ActionType `json:"type"`
	AttemptedAction string     `json:"attempted_action"`
}

// ActionRequest is the canonical request sent from enforcement points to the daemon.
type ActionRequest struct {
	RequestID   string       `json:"request_id"`
	Timestamp   string       `json:"timestamp"`
	Actor       Actor        `json:"actor"`
	Environment Environment  `json:"environment"`
	Action      ActionDetail `json:"action"`
	Resource    Resource     `json:"resource"`
}

// ValidateActionRequest returns a list of validation errors, or nil if valid.
func ValidateActionRequest(req *ActionRequest) []string {
	var errs []string
	if req.RequestID == "" {
		errs = append(errs, "request_id is required")
	}
	if req.Timestamp == "" {
		errs = append(errs, "timestamp is required")
	}
	if req.Actor.UserID == "" {
		errs = append(errs, "actor.user_id is required")
	}
	if req.Actor.AgentType == "" {
		errs = append(errs, "actor.agent_type is required")
	}
	if req.Action.Type == "" {
		errs = append(errs, "action.type is required")
	}
	if req.Action.AttemptedAction == "" {
		errs = append(errs, "action.attempted_action is required")
	}
	if req.Resource.Kind == "" {
		errs = append(errs, "resource.kind is required")
	}
	return errs
}

File: types/action.ts
/**
 * Author: Deepankar Das
 */

/**
 * Enforcer — Canonical Action Types
 *
 * Defines the ActionRequest contract used by all enforcement points
 * to submit actions to the daemon for policy evaluation.
 *
 * TDD Reference: Appendix B.4 (ActionRequest)
 * PRD Reference: Appendix C, R-1
 */

import { z } from "zod";

// ── Action Types ────────────────────────────────────────────────────────────

export const ActionType = z.enum([
  "file.read",
  "file.write",
  "file.delete",
  "file.move",
  "shell.exec",
  "network.request",
  // Phase 2 surfaces:
  "package.install",
  "credential.access",
  "mcp.invoke",
]);
export type ActionType = z.infer<typeof ActionType>;

// ── Resource Kind ───────────────────────────────────────────────────────────

export const ResourceKind = z.enum([
  "file",
  "command",
  "host",
  "mcp_tool",
  "credential",
  "database",
]);
export type ResourceKind = z.infer<typeof ResourceKind>;

// ── Resource Classification Tags ────────────────────────────────────────────

export const ResourceClassification = z.enum([
  "destructive",
  "network_tool",
  "package_manager",
  "sensitive_path",
  "safe",
  "potential_exfiltration",
  "bypass_attempt",
]);
export type ResourceClassification = z.infer<typeof ResourceClassification>;

// ── Actor ───────────────────────────────────────────────────────────────────

export const ActorSchema = z.object({
  user_id: z.string().min(1),
  agent_type: z.string().min(1),       // "claude_code" | "cursor" | "codex"
  agent_instance: z.string().min(1),    // "vscode_ext_1" | "cli_session_2"
  session_id: z.string().min(1),
});
export type Actor = z.infer<typeof ActorSchema>;

// ── Environment ─────────────────────────────────────────────────────────────

export const EnvironmentSchema = z.object({
  workspace: z.string().min(1),         // Absolute path to project root
  repo: z.string().default(""),         // "org/repo-name"
  branch: z.string().default(""),       // "main", "feature/x"
  tier: z.string().default("development"), // "development" | "staging" | "production"
  deployment_mode: z.string().default("host"), // "host" | "container" | "remote"
});
export type Environment = z.infer<typeof EnvironmentSchema>;

// ── Resource ────────────────────────────────────────────────────────────────

export const ResourceSchema = z.object({
  kind: ResourceKind,
  path: z.string().optional(),           // File path (for file ops)
  host: z.string().optional(),           // Destination host (for network ops)
  value: z.string().optional(),          // Command string, URL, etc.
  classification: z.array(ResourceClassification).default([]),
});
export type Resource = z.infer<typeof ResourceSchema>;

// ── Action ──────────────────────────────────────────────────────────────────

export const ActionDetailSchema = z.object({
  type: ActionType,
  attempted_action: z.string().min(1),   // Human-readable: "Write 247 bytes to ~/.config/settings.json"
});
export type ActionDetail = z.infer<typeof ActionDetailSchema>;

// ── ActionRequest (full contract) ───────────────────────────────────────────

export const ActionRequestSchema = z.object({
  request_id: z.string().min(1),
  timestamp: z.string().datetime(),
  actor: ActorSchema,
  environment: EnvironmentSchema,
  action: ActionDetailSchema,
  resource: ResourceSchema,
});
export type ActionRequest = z.infer<typeof ActionRequestSchema>;